// bunnycdn-node/dist/lib/models/pullzone.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var decorators_1 = require("class-transformer/decorators");
var host_1 = require("./host");
var PullZone = /** @class */ (function () {
    function PullZone(name, originUrl) {
        this.name = name;
        this.originUrl = originUrl;
        this.id = -1;
        this.enabled = false;
        this.hostNames = [];
        this.storageZoneId = -1;
        this.allowedReferrers = [];
        this.blockedIps = [];
        this.enableGeoZoneUS = false;
        this.enableGeoZoneEU = false;
        this.zoneSecurityEnabled = false;
        this.zoneSecurityKey = '';
        this.zoneSecurityIncludeHashRemoteIP = false;
        this.ignoreQueryStrings = false;
        this.monthlyBandwidthLimit = -1;
        this.monthlyBandwidthUsed = -1;
        this.addHostHeader = false;
        this.zoneType = -1;
        this.accessControlOriginHeaderExtensions = [];
        this.enableAccessControlOriginHeader = false;
        this.disabledCookies = false;
        this.budgetRedirectedCountries = [];
        this.blockedCountries = [];
    }
    __decorate([
        decorators_1.Type(function () { return host_1.Host; }),
        __metadata("design:type", Array)
    ], PullZone.prototype, "hostNames", void 0);
    return PullZone;
}());
exports.PullZone = PullZone;
