// bunnycdn-node/dist/lib/bunny.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
// @format
require("reflect-metadata");
var class_transformer_1 = require("class-transformer");
var axios_1 = require("axios");
var clients_1 = require("./clients");
var models_1 = require("./models");
var enums_1 = require("./enums");
var Bunny = /** @class */ (function (_super) {
    __extends(Bunny, _super);
    function Bunny(httpClient) {
        var _this = _super.call(this, httpClient || axios_1.default) || this;
        _this._pullZoneClient = new clients_1.PullZoneClient(httpClient || axios_1.default);
        _this._storageClient = new clients_1.StorageClient(httpClient || axios_1.default);
        return _this;
    }
    Object.defineProperty(Bunny.prototype, "pullZone", {
        get: function () {
            return this._pullZoneClient;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Bunny.prototype, "storage", {
        get: function () {
            return this._storageClient;
        },
        enumerable: true,
        configurable: true
    });
    Bunny.prototype.statistics = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _super.prototype.fetch.call(this, enums_1.Resource.Statistic)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, class_transformer_1.plainToClass(models_1.Statistic, response.data)];
                }
            });
        });
    };
    Bunny.prototype.billing = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _super.prototype.fetch.call(this, enums_1.Resource.Billing)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, class_transformer_1.plainToClass(models_1.Billing, response.data)];
                }
            });
        });
    };
    Bunny.prototype.applyCode = function (couponCode) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _super.prototype.fetch.call(this, enums_1.Resource.Billing + "/" + enums_1.Endpoint.ApplyCode + "?" + enums_1.Parameter.CouponCode + "=" + couponCode)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    Bunny.prototype.purge = function (url) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _super.prototype.post.call(this, enums_1.Endpoint.Purge + "?" + enums_1.Parameter.URL + "=" + url, null)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    Bunny.prototype.hardUpdate = function (host, storageZone, path, fileName, fileContents) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.update(storageZone + "/" + path + "/" + fileName, fileContents)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.purge(host + "/" + path + "/" + fileName)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return Bunny;
}(clients_1.HttpBase));
exports.Bunny = Bunny;
